### PR TITLE
treedb: apply r3a create collection slice

### LINKED ISSUE #3041: TreeDB Raft R3a-3/R3a-4: create collection slice and logical digest
## Goal

Land the first R3a vertical slice: create collection applies through local command WAL and normal catalog executors, with `LogicalDigestV1` convergence proof.

Done means the same deterministic create-collection entry sequence applied to two fresh DBs produces deterministic results and matching logical digests without relying on byte-identical local files, page IDs, root IDs, value-log RIDs, or local WAL LSNs.

## Parent

Parent tracker: #1654

## Scope

- [ ] Implement create collection as the first accepted command.
- [ ] Lower create collection entries to `CatalogCreateCollection` command-WAL payloads.
- [ ] Define deterministic collection identity/generation behavior before any data-mutation widening.
- [ ] Apply through local command WAL and normal catalog executor path.
- [ ] Define `LogicalDigestV1` for supported catalog state.
- [ ] Prove checkpoint/reopen and local storage layout independence.

## Commands In Scope

- create collection only

## Commands Explicitly Out Of Scope Or Rejected

- insert, replace/update, delete
- create/drop index
- drop collection
- flush/checkpoint as replicated commands
- networked Raft

## Required Tests

- [ ] Create collection apply produces expected catalog state and deterministic `ApplyResultV1`.
- [ ] Duplicate create with same identity/metadata returns defined same-result or already-applied behavior.
- [ ] Duplicate create with same name but incompatible identity/metadata fails before incompatible mutation.
- [ ] Reopen after create preserves state and progress rules.
- [ ] Two fresh DBs applying the same create sequence produce the same `LogicalDigestV1`.
- [ ] Different catalog or target identity changes the digest.
- [ ] Different checkpoint timing does not change the digest.
- [ ] Digest excludes page IDs, root IDs, local WAL LSN, segment names, local value-log RIDs, and file paths.

## Benchmarks And Evidence

- [ ] Benchmark create-collection decode/lower/apply if this path adds measurable overhead.
- [ ] Report entries/s or commands/s, `ns/op`, `B/op`, `allocs/op`, command-WAL frames appended, and digest computation cost where applicable.
- [ ] Before/after comparison is required for optimization claims.

## Branch And PR Policy

- [ ] Topic branch from current `main`.
- [ ] PR body includes start-phase tests, close-phase tests, benchmark evidence or rationale, and #1654 linkage.
- [ ] Latest-head CI is required; stale green checks do not count.
- [ ] AI reviews are requested only after coherent code and current evidence exist.

## Completion Criteria

- [ ] Create collection is the only accepted mutation in this slice.
- [ ] Deterministic identity/generation is resolved or this issue remains blocked.
- [ ] `LogicalDigestV1` is strong enough for later insert/replace/delete convergence tests.

### PR DESCRIPTION
**Pull request by snissn**
Refs #1654.
Closes #3041.
Supersedes draft #3062 with a clean branch rebased onto merged #3040.

## Summary
- Adds the create-collection R3a apply slice through local command-WAL append/finalize and the normal collections catalog executor.
- Adds LogicalDigestV1 over supported catalog state so fresh DBs can prove deterministic convergence without byte-identical local files, roots, page IDs, value-log RIDs, or WAL LSNs.
- Preserves the #3040 apply rejection boundary while widening only CatalogCreateCollection as the first accepted mutation.

## Validation
- GOWORK=off go test ./TreeDB/internal/raftapply ./TreeDB/internal/raftentry ./TreeDB/internal/commandwalapply ./TreeDB/internal/nativewire -count=1 -timeout=3m
- GOWORK=off go test -count=1 ./TreeDB/collections -run 'TestCollectionCommandWALCreateCollection' -timeout=3m\n- GOWORK=off go test -count=1 ./TreeDB/db -run 'TestTrustedCommandWALIntentAppendsCanonicalCollectionPayload|TestCommandWALAppliedCommandLSNMetaFieldRoundTrip' -timeout=3m\n- GOWORK=off go vet ./TreeDB/internal/raftapply ./TreeDB/internal/raftentry ./TreeDB/internal/commandwalapply ./TreeDB/internal/nativewire\n- git diff --check origin/main...HEAD\n\n## Benchmarks\nNo benchmark run: this slice is correctness scaffolding for one catalog command and digest proof, with no optimization claim.

## Summary by CodeRabbit

* **New Features**
  * Added new collection-creation entrypoints that support explicit command-log intent injection and prepared intent workflows (with status/preflight variants), improving replay reliability.
  * Introduced deterministic replicated “create collection” application, including logical convergence hashing.
* **Bug Fixes**
  * Improved idempotency and already-applied handling to prevent duplicate work and inconsistent progress/result updates.
  * Fixed command-log noop publishing to respect staged intent and avoid unsafe locking behavior.
* **Tests**
  * Added coverage for prepared-intent locking, invalid intent rejection, and deterministic apply edge cases.

### CODE REVIEW

File: TreeDB/internal/raftapply/create_collection.go [hunk 1]
```diff
@@ -0,0 +1,363 @@
...
+	handle, _, err := h.walApply.Append(h.db, frame, commandwalapply.ApplyMetadata{}, commandwalapply.Options{Sync: meta.SyncLocalCommandWAL})
+	if err != nil {
+		return raftentry.ApplyResultV1{}, codeCommandWALApplyError(err)
+	}
+	intent := handle.CommandWALIntent()
+	if intent == nil || handle.LSN() == 0 {
+		return raftentry.ApplyResultV1{}, codedError(raftentry.ErrorUnsafeDurabilityModeV1, "raftapply: command WAL append did not return a usable intent")
+	}
+	if _, err := collections.NewCollectionManager(h.db).CreateCollectionWithCommandWALIntent(collectionMeta, intent); err != nil {
+		return raftentry.ApplyResultV1{}, codeCollectionApplyError(err)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go]: Fixed in fbdee1ee3: apply now aborts unreleased staged command-WAL handles on every non-finalized return, and the abort path poisons unrecovered appended frames while releasing the staging guard. Added commandwalapply coverage for the blocked public publish case.

File: TreeDB/internal/raftapply/create_collection.go [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+	}
+	if _, err := h.walApply.Finalize(h.db, handle, commandwalapply.ApplyMetadata{}, commandwalapply.Options{Sync: meta.SyncLocalCommandWAL}); err != nil {
+		return raftentry.ApplyResultV1{}, codeCommandWALApplyError(err)
+	}
+	logical, err := LogicalDigestV1ForDB(h.db, LogicalDigestOptionsV1{
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go]: Fixed in fbdee1ee3: the create-collection logical digest is computed after the catalog executor publishes the entry but before commandwalapply.Finalize releases the staging guard, so the stored result stays inside the staged apply boundary.

File: TreeDB/internal/raftapply/apply.go [hunk 3]
```diff
@@ -133,11 +126,48 @@ func (h *Harness) ApplyCommittedEntryV1(entryBytes []byte, meta ApplyMetadataV1)
 			if record.CommandDigest != entry.Digest {
 				return reject(entry.Digest, raftentry.ErrorRejectedConflictV1, fmt.Errorf("raftapply: apply entry %d/%d digest conflicts with existing result", meta.EntryID.Term, meta.EntryID.Index))
 			}
-			return reject(entry.Digest, raftentry.ErrorResultReplayRequiredV1, fmt.Errorf("raftapply: apply entry %d/%d result replay is not implemented", meta.EntryID.Term, meta.EntryID.Index))
+			return record.Result, nil
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/apply.go:168]: Fixed in fbdee1ee3: stored-result replay now records matching apply progress before returning the cached result, with coverage for a result-present/progress-missing retry.

File: TreeDB/internal/raftapply/create_collection.go [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+	defer func() {
+		if !handleFinalized {
+			h.walApply.Abort(h.db, handle)
+		}
+	}()
+	intent := handle.CommandWALIntent()
+	if intent == nil || handle.LSN() == 0 {
+		return raftentry.ApplyResultV1{}, codedError(raftentry.ErrorUnsafeDurabilityModeV1, "raftapply: command WAL append did not return a usable intent")
+	}
+	if _, err := collections.NewCollectionManager(h.db).CreateCollectionWithCommandWALIntent(collectionMeta, intent); err != nil {
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go]: Fixed in 131b41eac: raftapply now owns a replay-safe collection manager on the harness, and the create/preflight/logical-digest paths use `NewCommandWALReplayCollectionManager` instead of transient live managers. The replay manager is exported with coverage confirming it does not register backend raw-publish barriers or close hooks.

File: TreeDB/internal/raftapply/create_collection.go [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+	}
+	alreadyApplied, err := h.preflightCreateCollectionV1(collectionMeta, payload)
+	if err != nil {
+		return raftentry.ApplyResultV1{}, err
+	}
+	frame, err := commandwalapply.CatalogCreateCollectionFrame(payload)
+	if err != nil {
+		return raftentry.ApplyResultV1{}, codedError(raftentry.ErrorMalformedEntryV1, "raftapply: lower catalog create collection: %v", err)
+	}
+	handle, _, err := h.walApply.Append(h.db, frame, commandwalapply.ApplyMetadata{}, commandwalapply.Options{Sync: meta.SyncLocalCommandWAL})
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go]: Fixed in 3d0950647: create-collection apply now enters the collection catalog/schema executor before staging the command-WAL frame. The collection manager exposes a prepared-intent callback that runs under the schema lock, and raftapply appends through that callback, preserving the public create lock order. Added coverage proving the prepared callback runs while the schema lock is held.

File: TreeDB/internal/raftapply/create_collection.go [hunk 6]
```diff
@@ -0,0 +1,390 @@
...
+	switch value {
+	case 2:
+		return collections.IndexValueBool
+	case 3:
+		return collections.IndexValueInt64
+	case 4:
+		return collections.IndexValueDouble
+	default:
+		return collections.IndexValueString
+	}
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go]: Fixed in c04059216: the create metadata decoders now reject unsupported document format, root storage, index value type, and index storage policy enums with coded unsupported-feature errors. Added direct apply-decoder coverage for each enum boundary.

File: TreeDB/internal/raftapply/create_collection.go [hunk 7]
```diff
@@ -0,0 +1,416 @@
...
+
+func lowerCreateCollectionV1(entry raftentry.CommandEntryV1) (collections.CollectionMeta, []byte, error) {
+	if entry.Target.CommandID != nativewire.CommandCreateCollection {
+		return collections.CollectionMeta{}, nil, codedError(raftentry.ErrorUnsupportedCommandV1, "raftapply: unsupported command %d", entry.Target.CommandID)
+	}
+	meta, err := decodeCreateCollectionMetaV1(entry.Target.CollectionMeta)
+	if err != nil {
+		return collections.CollectionMeta{}, nil, err
+	}
+	payload, err := collections.EncodeCatalogCreateCollectionCommandWALPayload(meta)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go:146]: Fixed in 34210430f: create apply now decodes `expected_catalog_version` before lowering/publishing, requires caller-supplied current catalog metadata in `ApplyMetadataV1`, and rejects target-free stale creates as `rejected-conflict` before any command-WAL append. The collection executor also composes the same guard into the locked publish preflight while preserving the same-schema already-applied no-op path. Added `TestCreateCollectionStaleCatalogGuardFailsBeforeAppend` to assert no seam calls, no new WAL frame, unchanged `AppliedCommandLSN`, and no created collection.

File: TreeDB/internal/raftapply/create_collection.go [hunk 8]
```diff
@@ -0,0 +1,456 @@
...
+}
+
+func decodeCreateCollectionMetaV1(raw []byte) (collections.CollectionMeta, error) {
+	off := 0
+	version, err := readCreateMetaUvarint(raw, &off, "collection_meta.version")
+	if err != nil {
+		return collections.CollectionMeta{}, err
+	}
+	if version != 1 {
+		return collections.CollectionMeta{}, codedError(raftentry.ErrorUnsupportedVersionV1, "raftapply: collection_meta version %d", version)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go]: Fixed in 5acc83409: raftapply now accepts collection_meta versions 1..5, parses the current vector-index count/definition tail, and the internal deterministic-entry validator accepts the same current wire shape. Added coverage that builds a v5 create-collection entry through the registry and applies it successfully.

File: TreeDB/internal/raftapply/create_collection.go [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+
+func lowerCreateCollectionV1(entry raftentry.CommandEntryV1) (collections.CollectionMeta, []byte, error) {
+	if entry.Target.CommandID != nativewire.CommandCreateCollection {
+		return collections.CollectionMeta{}, nil, codedError(raftentry.ErrorUnsupportedCommandV1, "raftapply: unsupported command %d", entry.Target.CommandID)
+	}
+	meta, err := decodeCreateCollectionMetaV1(entry.Target.CollectionMeta)
+	if err != nil {
+		return collections.CollectionMeta{}, nil, err
+	}
+	payload, err := collections.EncodeCatalogCreateCollectionCommandWALPayload(meta)
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go:146]: Fixed in 5acc83409: create lowering now applies the nativewire storage-policy override before encoding the catalog command-WAL payload, forcing data root, index state, and scalar index storage to RootStorageFast. Added coverage that a compressed request stores fast policies in both the live catalog and command-WAL payload.

File: TreeDB/internal/raftapply/create_collection.go [hunk 10]
```diff
@@ -0,0 +1,456 @@
...
+	if !handleAppended {
+		return raftentry.ApplyResultV1{}, codedError(raftentry.ErrorUnsafeDurabilityModeV1, "raftapply: create collection did not append a command WAL frame")
+	}
+	logical, err := h.logicalDigestV1(LogicalDigestOptionsV1{
+		ScopeRule:     meta.ScopeRule,
+		DatabaseScope: meta.DatabaseScope,
+		CatalogScope:  meta.CatalogScope,
+	})
+	if err != nil {
+		return raftentry.ApplyResultV1{}, err
```
[the remark this answers is not in the document]
Comment (snissn) [reply TreeDB/internal/raftapply/create_collection.go]: Fixed in 7663674f: if logical digest computation fails after the catalog executor has published the create, apply now returns recovery-required instead of a deterministic rejection. Added coverage with an injected post-publish digest failure that verifies the collection remains visible and AppliedCommandLSN is advanced.
